pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/PinotDataBuffer.java

```diff
@@ -110,7 +110,8 @@ public String toString() {
   private static final AtomicLong MMAP_BUFFER_COUNT = new AtomicLong();
   private static final AtomicLong MMAP_BUFFER_USAGE = new AtomicLong();
   private static final AtomicLong ALLOCATION_FAILURE_COUNT = new AtomicLong();
-  private static final Map<PinotDataBuffer, BufferContext> BUFFER_CONTEXT_MAP = new WeakHashMap<>();
+  // we need to use MapMaker instead of WeakHashMap because we want to use identity comparison for the keys
+  private static final Map<PinotDataBuffer, BufferContext> BUFFER_CONTEXT_MAP = new MapMaker().weakKeys().makeMap();
```

Comment:
Seems like the returned map can be accessed concurrently. We might be able to remove the `synchronized` block